### PR TITLE
Instance: Sara's Memory

### PR DESCRIPTION
**Pull request by Ridley8819**
-- Quest DB
---- Sara's Memory
---- (+ Moved ID 15000 and 15001 to correct location)
-- Mob DB
---- 2542 - 2546
-- Instance Script
-- scripts.conf
---- Added Sara's Memory

Sources:
http://irowiki.org/wiki/Sara's_Memory
https://www.youtube.com/watch?v=5wLBuGp4CKA
https://www.youtube.com/watch?v=aMDLOEI0MBs
http://ro.vrc7.net/?page_id=1556
http://www.divine-pride.net/database/monster/2543 (and ongoing)
Original Script:
http://ziu-ro.blogspot.de/2012/10/heroes-trails-part-i-saras-memories.html

---
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1383)

### CODE REVIEW

File: npc/re/instances/Sara'sMemory.txt [hunk 1]
```diff
@@ -0,0 +1,1863 @@
...
+			.@p_name$ = getpartyname(.@party_id);
+			.@md_name$ = "Sara's Memory";
+			if (!instance_check_party(.@party_id)) {
+				mes("[Leon the Adventurer]");
+				mes("Before you enter you need to organize a party! Don't know how? Type /organize PARTYNAMEHERE. You\r"
+					"can use quotes to put spaces in a party name.");
+				close;
+			}
+			if (getcharid(0) == getpartyleader(.@party_id, 2))
+				.@menu1$ = "Boot up the dimensional device";
```
Comment (Asheraf) [inline npc/re/instances/Sara'sMemory.txt]: maybe an additional check here
`if (has_instance("1@sara"))`
to remove the choise instead of show an error ? or it's officialy like this?

**Comment by MishimaHaruna** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[npc/re/scripts.conf, line 94 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwM0lC5gesOd6vPI1M:-KNwM0lDsHdQXUPA2CjE:-662876387) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/scripts.conf#L94)):*
> ```
> npc: npc/re/instances/OctopusCave.txt
> npc: npc/re/instances/OldGlastHeim.txt
> npc: npc/re/instances/Sara'sMemory.txt
> ```

Can we limit the file names to letters, numbers, dashes, underscores, please? (my personal preference being lowercase letters, numbers, underscores, but uppercase letters and dashes are also okay)

---

*[npc/re/instances/Sara'sMemory.txt, line 43 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwMlPMSprXxeoTyiNX:-KNwMlPMSprXxeoTyiNY:-1924808812) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L43)):*
> ```
> 
> dali,130,107,5	script	Leon the Adventurer#Sara	4_M_DST_GRAND,{ // instance CD check only at the Dimensional Device
> 		mes("[Leon the Adventurer]");
> ```

Could you re-check the indentation here? (it makes it a bit hard to follow the script flow, when the indentation changes randomly)

---

*[npc/re/instances/Sara'sMemory.txt, line 86 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwN7kqMikIci2I7cMs:-KNwN7kqMikIci2I7cMt:-2040653694) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L86)):*
> ```
> 						if (select("Yes","No") == 2)
> 							close;
> 					next;
> ```

Is this intended? (the `next` after the `if`, instead of before)

---

*[npc/re/instances/Sara'sMemory.txt, line 101 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwNScNOuCqgnD3-KPp:-KNwNScOcJ4E6nRclYiQ:-1187757238) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L101)):*
> ```
> 						close;
> 				switch (select("Okay", "Give up")) {
> 					next;
> ```

You can't put a `next` inside a `switch`, but outside of any `case`

---

*[npc/re/instances/Sara'sMemory.txt, line 153 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwNffB5Z1c53JqURR3:-KNwNffCkHus3RSfqRA4:1213800876) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L153)):*
> ```
> 				close;
> 			}
> 		}
> ```

Yep, indentation completely messed up (if indentation is correct, the final `}` would be at column 0)

---

*[npc/re/instances/Sara'sMemory.txt, line 1863 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwOnitj2pHzmXnknYP:-KNwOniuDkF58eI_NVF5:1791277930) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L1863)):*
> ```
> 	disablenpc instance_npcname("Sara Irene#sarains23");
> 	end;
> }
> ```

The script is missing a newline here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1383)*

**Comment by Ridley8819** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[npc/re/scripts.conf, line 94 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwM0lC5gesOd6vPI1M:-KO1WzyQ-AYvWpmLAME5:1763312785) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/scripts.conf#L94)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Can we limit the file names to letters, numbers, dashes, underscores, please? (my personal preference being lowercase letters, numbers, underscores, but uppercase letters and dashes are also okay)
</details>
What would you prefer then? I followed here the herc  structure (personally i preffer lowercase too, but we would have to rename all the instances then, else it looks aweful). I then would say to make it saras_memory (old_glast_heim, octopus_cave, ...)

---

*[npc/re/instances/Sara'sMemory.txt, line 43 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwMlPMSprXxeoTyiNX:-KO1XXksK34NhOEKtLYA:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L43)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Could you re-check the indentation here? (it makes it a bit hard to follow the script flow, when the indentation changes randomly)
</details>
Done.

---

*[npc/re/instances/Sara'sMemory.txt, line 86 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwN7kqMikIci2I7cMs:-KO1_GSbPbuzAp63bRhd:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L86)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Is this intended? (the `next` after the `if`, instead of before)
</details>
Done.

---

*[npc/re/instances/Sara'sMemory.txt, line 101 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwNScNOuCqgnD3-KPp:-KO1_Gwb3tnr1opKHRc9:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L101)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> You can't put a `next` inside a `switch`, but outside of any `case`
</details>
Done.

---

*[npc/re/instances/Sara'sMemory.txt, line 153 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwNffB5Z1c53JqURR3:-KO1_HfF1svdmUAJKIky:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L153)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Yep, indentation completely messed up (if indentation is correct, the final `}` would be at column 0)
</details>
Done.

---

*[npc/re/instances/Sara'sMemory.txt, line 181 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNpliUenQkP2LTSkUll-r1-181:-KO1_IXo9KY-hwHS-g3n:-1825487674) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L181)):*
<details><summary><i>Previously, Asheraf (Acheraf Rian) wrote…</i></summary>
> maybe an additional check here
> 
> `if (has_instance("1@sara"))`
> 
> to remove the choise instead of show an error ? or it's officialy like this?
</details>
As you say, it's intended like this

---

*[npc/re/instances/Sara'sMemory.txt, line 1863 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwOnitj2pHzmXnknYP:-KO1_Ldc6jqHBsrY6Ny5:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/Sara'sMemory.txt#L1863)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> The script is missing a newline here
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1383)*

**Comment by MishimaHaruna** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[npc/re/scripts.conf, line 94 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwM0lC5gesOd6vPI1M:-KO24tEb-_SRGPPGJ11h:796257935) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/scripts.conf#L94)):*
<details><summary><i>Previously, Ridley8819 (Ridley) wrote…</i></summary>
> What would you prefer then? I followed here the herc  structure (personally i preffer lowercase too, but we would have to rename all the instances then, else it looks aweful). I then would say to make it saras_memory (old_glast_heim, octopus_cave, ...)
</details>
Up to you. The only problem was the apostrophe (I'd like to avoid special characters in file names). The way you named it now looks perfect to me (yup, I see your point about the other scripts' names, but I have to admit I never liked camelcase)

---

*[npc/re/instances/saras_memory.txt, line 98 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KO25_3JZ8nTj3tPXCHq:-KO25_3JZ8nTj3tPXCHr:-1405991104) ([raw file](https://github.com/herculesws/hercules/blob/20d4fb2b00a85f54bab8025e459c73a34b4e818e/npc/re/instances/saras_memory.txt#L98)):*
> ```
> 	if (questprogress(15003) == 1) {
> 		mes("I am not sure you have fully explored the crack in space and time yet. Perhaps you should look a bit further into it.");
> 		next;
> ```

This should probably be after the `if (!sarainstance) close;` bit, so that it shows *either* the close *or* the next (rather than show the next, then try to replace it with close, which I'm not 100% sure the client will accept)

---

*[npc/re/instances/saras_memory.txt, line 210 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KO26I_NDfPmomUigOYL:-KO26I_OeuUDIU0FIr2G:1509232767) ([raw file](https://github.com/herculesws/hercules/blob/20d4fb2b00a85f54bab8025e459c73a34b4e818e/npc/re/instances/saras_memory.txt#L210)):*
> ```
> 				close;
> 			}
> 			mapannounce("dali", sprintf(_$("%s, member of the party %s entered the instance %s."), strcharinfo(0), .@p_name$, .@md_name$), bc_map, C_SPRINGGREEN);
> ```

C_SPRINGGREEN should be added to constants.conf as well (I believe it's still not there)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1383)*

**Comment by Ridley8819** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[npc/re/scripts.conf, line 94 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNwM0lC5gesOd6vPI1M:-KO4BFroFbzQsd41cssK:545348909) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/scripts.conf#L94)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Up to you. The only problem was the apostrophe (I'd like to avoid special characters in file names). The way you named it now looks perfect to me (yup, I see your point about the other scripts' names, but I have to admit I never liked camelcase)
</details>
I do pr's then for the others 

---

*[npc/re/instances/saras_memory.txt, line 210 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KO26I_NDfPmomUigOYL:-KO4B5a0MuEDkQ2qVRSk:-814210196) ([raw file](https://github.com/herculesws/hercules/blob/20d4fb2b00a85f54bab8025e459c73a34b4e818e/npc/re/instances/saras_memory.txt#L210)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> C_SPRINGGREEN should be added to constants.conf as well (I believe it's still not there)
</details>
https://github.com/HerculesWS/Hercules/pull/1381
This is done here already

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1383)*

**Comment by Ridley8819** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[npc/re/instances/saras_memory.txt, line 181 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KNpliUenQkP2LTSkUll-r1-181:-KO5Z8cR8RsIPGyzioes:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/da803c1fd7cd2cdcbc99eafb7168288ca43bb2b7/npc/re/instances/saras_memory.txt#L181)):*
<details><summary><i>Previously, Ridley8819 (Ridley) wrote…</i></summary>
> As you say, it's intended like this
</details>
Done.

---

*[npc/re/instances/saras_memory.txt, line 98 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KO25_3JZ8nTj3tPXCHq:-KO5YwxqSe_n9WTJancj:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/20d4fb2b00a85f54bab8025e459c73a34b4e818e/npc/re/instances/saras_memory.txt#L98)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> This should probably be after the `if (!sarainstance) close;` bit, so that it shows *either* the close *or* the next (rather than show the next, then try to replace it with close, which I'm not 100% sure the client will accept)
</details>
Done.

---

*[npc/re/instances/saras_memory.txt, line 210 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1383#-KO26I_NDfPmomUigOYL:-KO5Z0avy7JAyB2lUhtp:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/20d4fb2b00a85f54bab8025e459c73a34b4e818e/npc/re/instances/saras_memory.txt#L210)):*
<details><summary><i>Previously, Ridley8819 (Ridley) wrote…</i></summary>
> https://github.com/HerculesWS/Hercules/pull/1381
> This is done here already
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1383)*
